### PR TITLE
chore(python): update sift benchmarks to include GIST1M as well

### PR DESCRIPTION
**Pull request by changhiskhan**
Previously our benchmarks were hard coded for SIFT (128D). This PR does a few things:

1. Allow the user to parametrize the number of dimensions to handle GIST1M (960D)
2. Support external query vectors (e.g., GIST1M provides a 1000 vector subset for querying).

To run the benchmarks, use the scripts in this directory (`/lance/benchmarks/sift`):

1. Download the `gist.tar.gz` from http://corpus-texmex.irisa.fr/ and unzip
2. Generate Lance datasets using the `datagen.py` script:
- Database vectors: `./datagen.py ~/Downloads/gist/gist_base.fvecs ~/.lancedb/gist1m.lance -g 1024 -m 50000 -d 960`
- Query vectors: `./datagen.py ~/Downloads/gist/gist_query.fvecs ~/.lancedb/gist_query.lance -g 1024 -m 50000 -d 960 -n 1000`
3. Create index: `./index.py ~/.lancedb/gist1m.lance -i 256 -p 120` (`-i` is ivf partitions and `-p` is pq subvectors)
4. Run the benchmark: `./metrics.py ~/.lancedb/gist1m.lance results.csv -i 256 -p 120 -q ~/.lancedb/gist_query.lance -k 1` (`-k` is how many results to fetch  and `-q` is the query vectors)

image

### CODE REVIEW

File: benchmarks/sift/datagen.py [hunk 1]
```diff
@@ -31,23 +31,40 @@ def main():
...
+        help="Number of vectors to generate",
+        metavar="NUM",
+    )
     args = parser.parse_args()
 
     with open(args.data, mode="rb") as fobj:
         buf = fobj.read()
-        data = np.array(struct.unpack("<128000000f", buf[4 : 4 + 4 * 1000000 * 128]))
+        data = np.array(struct.unpack(f"<{args.dimension * args.num_vectors}f", 
+                                      buf[4 : 4 + 4 * args.num_vectors * args.dimension]))
```
Comment (QianZhu) [inline benchmarks/sift/datagen.py:55]: guess for the benchmark dataset size, we will not have any memoryerror concern

Comment (changhiskhan) [reply benchmarks/sift/datagen.py:55]: no it's only 1M vectors

**Review by QianZhu — APPROVED** [review]

File: benchmarks/sift/datagen.py (hunk 1 above)
Comment (prrao87) [reply benchmarks/sift/datagen.py:55]: Most public benchmarks are of this size and dimensionality, so it shouldn't be a concern as long as this code is used for those benchmarks.

**Comment by prrao87** [comment]
@changhiskhan I updated the docs to show how to run the SIFT/GIST-1M benchmarks. I think it also makes sense to backlink these benchmarks from the LanceDB repo too? It doesn't make sense to have two separate sets of benchmarks in either repo to maintain over time.

**Review by prrao87 — APPROVED** [review]
I ran both SIFT-1M and GIST-1M benchmarks on my 2022 M2 Macbook Pro

* <1 ms for SIFT-1M `ivf=256 pq=16 nprobes=10`
* <3 ms for GIST-1M `ivf=256 pq=120 nprobes=10`

Fixed the command line args for both as well.

**Pull request merged**